File: packages/python/secrefs/providers/vault.py
"""
HashiCorp Vault provider supporting both KV v1 and KV v2 secrets engines.
Auth is ambient via VAULT_ADDR/VAULT_TOKEN - point `path` at whatever the
Vault HTTP API itself expects (KV v2 mounts include a literal `data/`
segment, e.g. `secret/data/stripe`; KV v1 mounts do not).

hvac is synchronous, so calls are offloaded to a thread via
`asyncio.to_thread`. The client is constructed lazily on first use so that
simply having a VaultProvider in your registry doesn't require Vault to be
configured if you never reference sec://vault/... at all.

Read secrets are re-read on every expansion by default; see `cache_ttl_ms`
and ../ttl_cache.py for why that default is what it is.
"""

from __future__ import annotations

import asyncio
import json
import os
from typing import Any, Dict, Optional, cast

import hvac

from ..ttl_cache import TtlCache
from .base import ProviderHealth, SecretFetchRequest, SecretProvider, extract_field


class VaultProvider(SecretProvider):
    name = "vault"

    def __init__(
        self,
        url: Optional[str] = None,
        token: Optional[str] = None,
        client: Optional[Any] = None,
        cache_ttl_ms: float = 0.0,
    ) -> None:
        """`cache_ttl_ms` is how long a read secret may be reused; it
        defaults to 0, meaning every expansion re-reads, so rotation reaches
        a long-running consumer without a redeploy."""
        self._explicit_client = client
        self._url = url or os.environ.get("VAULT_ADDR")
        self._token = token or os.environ.get("VAULT_TOKEN")
        self._client_instance: Optional[Any] = None
        self._data_cache: TtlCache[Dict[str, Any]] = TtlCache(ttl_ms=cache_ttl_ms)

    def _get_client(self) -> Any:
        if self._explicit_client is not None:
            return self._explicit_client
        if self._client_instance is not None:
            return self._client_instance

        if not self._url:
            raise ValueError("VAULT_ADDR is not set (required for sec://vault/... references)")
        if not self._token:
            raise ValueError("VAULT_TOKEN is not set (required for sec://vault/... references)")

        self._client_instance = hvac.Client(url=self._url, token=self._token)
        return self._client_instance

    async def _fetch_data(self, path: str) -> Dict[str, Any]:
        # Outside the try: an unconfigured VAULT_ADDR/VAULT_TOKEN is a
        # misconfiguration, not a failed read, and shouldn't be reported as
        # one.
        client = self._get_client()
        try:
            response = await asyncio.to_thread(client.read, path)
        except Exception as exc:  # noqa: BLE001
            raise ValueError(f'could not read Vault path "{path}": {exc}') from exc

        if response is None or "data" not in response:
            raise ValueError(f'no data returned for path "{path}"')

        outer = response["data"]
        # KV v2 responses nest the secret under data.data alongside
        # data.metadata; KV v1 responses put the secret straight in data.
        if isinstance(outer, dict) and "data" in outer and "metadata" in outer:
            return cast(Dict[str, Any], outer["data"])
        return cast(Dict[str, Any], outer)

    async def fetch_one(self, request: SecretFetchRequest) -> str:
        data = await self._data_cache.fetch(request.path, lambda: self._fetch_data(request.path))

        if not request.field:
            if len(data) == 1:
                (only,) = data.values()
                return only if isinstance(only, str) else json.dumps(only)
            return json.dumps(data)

        return extract_field(json.dumps(data), request.field, provider=self.name, path=request.path)

    async def health_check(self) -> ProviderHealth:
        try:
            client = self._get_client()
            await asyncio.to_thread(client.sys.read_health_status)
            return ProviderHealth(provider=self.name, ok=True)
        except Exception as exc:  # noqa: BLE001
            return ProviderHealth(provider=self.name, ok=False, message=str(exc))

File: packages/node/src/providers/vault.ts
import vaultFactory from "node-vault";
import { TtlCache } from "../ttlCache.js";
import {
  BaseSecretProvider,
  errorMessage,
  extractField,
  type ProviderHealth,
  type SecretFetchRequest,
} from "./base.js";

export interface VaultProviderOptions {
  /** Defaults to $VAULT_ADDR. */
  endpoint?: string;
  /** Defaults to $VAULT_TOKEN. */
  token?: string;
  /** Inject a pre-configured client (primarily for testing). */
  client?: ReturnType<typeof vaultFactory>;
  /** How long a fetched secret may be reused, in ms. Defaults to 0 -
   * every expansion re-fetches, so rotation reaches a long-running
   * consumer without a redeploy. See ../ttlCache.ts. */
  cacheTtlMs?: number;
}

/**
 * HashiCorp Vault provider supporting both KV v1 and KV v2 secrets engines.
 * Auth is ambient via `VAULT_ADDR`/`VAULT_TOKEN` - point `path` at whatever
 * the Vault HTTP API itself expects (KV v2 mounts include a literal `data/`
 * segment, e.g. `secret/data/stripe`; KV v1 mounts do not).
 *
 * The client is constructed lazily on first use so that simply having a
 * `VaultProvider` in your provider registry doesn't require Vault to be
 * configured if you never actually reference `sec://vault/...`.
 */
export class VaultProvider extends BaseSecretProvider {
  readonly name = "vault";

  private readonly explicitClient?: ReturnType<typeof vaultFactory>;
  private readonly endpoint?: string;
  private readonly token?: string;
  private client: ReturnType<typeof vaultFactory> | null = null;
  private readonly dataCache: TtlCache<Record<string, unknown>>;

  constructor(options: VaultProviderOptions = {}) {
    super();
    this.explicitClient = options.client;
    this.endpoint = options.endpoint ?? process.env.VAULT_ADDR;
    this.token = options.token ?? process.env.VAULT_TOKEN;
    this.dataCache = new TtlCache<Record<string, unknown>>({ ttlMs: options.cacheTtlMs });
  }

  private getClient(): ReturnType<typeof vaultFactory> {
    if (this.explicitClient) return this.explicitClient;
    if (this.client) return this.client;

    if (!this.endpoint) {
      throw new Error("VAULT_ADDR is not set (required for sec://vault/... references)");
    }
    if (!this.token) {
      throw new Error("VAULT_TOKEN is not set (required for sec://vault/... references)");
    }

    this.client = vaultFactory({ endpoint: this.endpoint, token: this.token });
    return this.client;
  }

  private getData(path: string): Promise<Record<string, unknown>> {
    return this.dataCache.fetch(path, () =>
      this.getClient()
      .read(path)
      .then((response) => {
        const outer = response.data;
        if (outer === undefined || outer === null) {
          throw new Error(`no data returned for path "${path}"`);
        }
        // KV v2 responses nest the secret under data.data alongside
        // data.metadata; KV v1 responses put the secret straight in data.
        if (
          typeof outer === "object" &&
          "data" in (outer as Record<string, unknown>) &&
          "metadata" in (outer as Record<string, unknown>)
        ) {
          return (outer as Record<string, unknown>).data as Record<string, unknown>;
        }
        return outer as Record<string, unknown>;
      })
      .catch((err: unknown) => {
        throw new Error(`could not read Vault path "${path}": ${errorMessage(err)}`);
      }),
    );
  }

  async fetchOne(request: SecretFetchRequest): Promise<string> {
    const data = await this.getData(request.path);

    if (!request.field) {
      const keys = Object.keys(data);
      // A single-key secret with no #field requested resolves to that
      // key's raw value directly (e.g. Vault's common { value: "..." }
      // convention); anything else is returned as a JSON blob.
      if (keys.length === 1) {
        const only = data[keys[0] as string];
        return typeof only === "string" ? only : JSON.stringify(only);
      }
      return JSON.stringify(data);
    }

    return extractField(JSON.stringify(data), request.field, {
      provider: this.name,
      path: request.path,
    });
  }

  async healthCheck(): Promise<ProviderHealth> {
    try {
      await this.getClient().health();
      return { provider: this.name, ok: true };
    } catch (err) {
      return { provider: this.name, ok: false, message: errorMessage(err) };
    }
  }
}
